node/src/components/chainspec_loader.rs

```diff
@@ -482,6 +540,85 @@ mod tests {
     use super::*;
     use crate::{testing::TestRng, types::chainspec::CHAINSPEC_NAME};
 
+    #[test]
+    fn correctly_detects_when_to_exit_for_upgrade() {
+        let mut rng = crate::new_rng();
+        const HEIGHT: u64 = 10;
+        const PROTOCOL_VERSION: ProtocolVersion = ProtocolVersion::V1_0_0;
+        const IS_SWITCH: bool = false;
+        const VERIFIABLE_CHUNKED_HASH_ACTIVATION: u64 = 10;
+
+        let highest_block = None;
+        let next_upgrade_activation_point = None;
+        assert!(!ChainspecLoader::should_exit_for_upgrade(
+            highest_block,
+            next_upgrade_activation_point
+        ),);
+
+        let highest_block = Some(Box::new(Block::random_with_specifics(
+            &mut rng,
+            EraId::from(2),
+            HEIGHT,
+            PROTOCOL_VERSION,
+            IS_SWITCH,
+            VERIFIABLE_CHUNKED_HASH_ACTIVATION.into(),
+        )));
+        let next_upgrade_activation_point = None;
+        assert!(!ChainspecLoader::should_exit_for_upgrade(
+            highest_block,
+            next_upgrade_activation_point
+        ));
+
+        let highest_block = None;
+        let next_upgrade_activation_point = Some(ActivationPoint::EraId(10.into()));
+        assert!(!ChainspecLoader::should_exit_for_upgrade(
+            highest_block,
+            next_upgrade_activation_point
+        ));
+
+        let highest_block = Some(Box::new(Block::random_with_specifics(
+            &mut rng,
+            EraId::from(2),
+            HEIGHT,
+            PROTOCOL_VERSION,
+            IS_SWITCH,
+            VERIFIABLE_CHUNKED_HASH_ACTIVATION.into(),
+        )));
+        let next_upgrade_activation_point = Some(ActivationPoint::EraId(3.into()));
+        assert!(!ChainspecLoader::should_exit_for_upgrade(
+            highest_block,
+            next_upgrade_activation_point
+        ));
+
+        let highest_block = Some(Box::new(Block::random_with_specifics(
+            &mut rng,
+            EraId::from(2),
+            HEIGHT,
+            PROTOCOL_VERSION,
+            IS_SWITCH,
+            VERIFIABLE_CHUNKED_HASH_ACTIVATION.into(),
+        )));
+        let next_upgrade_activation_point = Some(ActivationPoint::EraId(2.into()));
+        assert!(ChainspecLoader::should_exit_for_upgrade(
+            highest_block,
+            next_upgrade_activation_point
+        ));
+
+        let highest_block = Some(Box::new(Block::random_with_specifics(
+            &mut rng,
+            EraId::from(2),
+            HEIGHT,
+            PROTOCOL_VERSION,
+            IS_SWITCH,
+            VERIFIABLE_CHUNKED_HASH_ACTIVATION.into(),
+        )));
+        let next_upgrade_activation_point = Some(ActivationPoint::EraId(2.into()));
+        assert!(ChainspecLoader::should_exit_for_upgrade(
+            highest_block,
+            next_upgrade_activation_point
+        ));
```

Comment:
I don't see the difference to the previous case.

Reply:
Me neither - I'll update to check for the case of the highest block being the last switch block before the next upgrade activation point.

Reply:
Fixed in [148a037](https://github.com/casper-network/casper-node/pull/2619/commits/148a037bdc00338d76e60afd30ff1d4834a10b95).